File: parser/src/test/java/com/aaroncoplan/waterfall/parser/tests/FunctionImplementationParsingTests.java
package com.aaroncoplan.waterfall.parser.tests;

import org.junit.Test;

public class FunctionImplementationParsingTests {

    @Test
    public void noArgs() {
        final String code = "module m {\nfunc myFunc() {}\n}";
        TestUtils.shouldPass(code);
    }

    @Test
    public void singleArg() {
        final String code = "module m {\nfunc myFunc(type1 arg1) {}\n}";
        TestUtils.shouldPass(code);
    }

    @Test
    public void multipleArgs() {
        final String[] code = {
            "module m {\nfunc myFunc(type1 arg1, type2 arg2) {}\n}",
            "module m {\nfunc myFunc(type1 arg1, type2 arg2, type3 arg3) {}\n}"
        };
        TestUtils.shouldPass(code);
    }

    @Test
    public void withReturnType() {
        final String[] code = {
            "module m {\nfunc myFunc(type1 arg1) returns type2 {}\n}",
            "module m {\nfunc myFunc() returns type2 {}\n}",
            "module m {\nfunc myFunc(type1 arg1, type4 arg2) returns type2 {}\n}"
        };
        TestUtils.shouldPass(code);
    }
}

File: parser/src/test/kotlin/com/aaroncoplan/waterfall/parser/tests/FunctionImplementationParsingTests.kt
package com.aaroncoplan.waterfall.parser.tests

import org.junit.Test

class FunctionImplementationParsingTests {

    @Test
    fun noArgs() {
        TestUtils.shouldPass("module m {\nfunc myFunc() {}\n}")
    }

    @Test
    fun singleArg() {
        TestUtils.shouldPass("module m {\nfunc myFunc(type1 arg1) {}\n}")
    }

    @Test
    fun multipleArgs() {
        TestUtils.shouldPass(arrayOf(
            "module m {\nfunc myFunc(type1 arg1, type2 arg2) {}\n}",
            "module m {\nfunc myFunc(type1 arg1, type2 arg2, type3 arg3) {}\n}"
        ))
    }

    @Test
    fun withReturnType() {
        TestUtils.shouldPass(arrayOf(
            "module m {\nfunc myFunc(type1 arg1) returns type2 {}\n}",
            "module m {\nfunc myFunc() returns type2 {}\n}",
            "module m {\nfunc myFunc(type1 arg1, type4 arg2) returns type2 {}\n}"
        ))
    }

    @Test
    fun withBody() {
        TestUtils.shouldPass("module m {\nfunc myFunc(type1 arg1) returns type2 {\n x := 5\n}\n}")
    }
}
